internal/application/documentation.go

```diff
@@ -0,0 +1,200 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	redskyappsv1alpha1 "github.com/thestormforge/optimize-controller/api/apps/v1alpha1"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// headComments is the mapping of field names to desired comments.
+var headComments = map[string]string{
+
+	"resources": `
+Resources define where you application's Kubernetes resources come from. These can be URL-like
+values such as file paths, HTTP URLs, or Git repository URLs. They can also be more complex
+definitions such references to in-cluster objects or Helm charts.
+`,
+
+	"parameters": `
+Parameters control what parts of you application will be optimized.
+
+Reference: https://docs.stormforge.io/reference/application/v1alpha1/#parameters
+`,
+
+	"scenarios": `
+Scenarios determine how you application will be put under load during optimization.
+
+Reference: https://docs.stormforge.io/reference/application/v1alpha1/#scenario
+`,
+
+	"objectives": `
+Objectives are used to define what you are trying to optimize about your application. Most
+objectives correspond to metrics observed over the course of an observation trial, for
+example: "p95-latency".
+
+Reference: https://docs.stormforge.io/reference/application/v1alpha1/#objective
+`,
+}
+
+// DocumentationFilter looks for Application instances and attempts to annotate them
+// with comments that might help people finish writing their app.yaml.
+type DocumentationFilter struct {
+	// Flag to completely disable documentation.
+	Disabled bool
+}
+
+// Filter applies documentation to any applications in the supplied node set.
+func (f *DocumentationFilter) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	if f.Disabled {
+		return nodes, nil
+	}
+
+	for _, n := range nodes {
+		meta, err := n.GetMeta()
+		if err != nil {
+			continue
+		}
+
+		if meta.Kind != "Application" || meta.APIVersion != redskyappsv1alpha1.GroupVersion.String() {
+			continue
+		}
+
+		if err := f.annotateApplication(n); err != nil {
+			return nil, err
+		}
+	}
+
+	return nodes, nil
+}
+
+// annotateApplication documents a single application.
+func (f *DocumentationFilter) annotateApplication(app *yaml.RNode) error {
+	n := app.YNode()
+
+	// Reminder about how the file was generated
+	if len(os.Args) >= 3 && filepath.Base(os.Args[0]) == "redskyctl" {
+		n.HeadComment = strings.Join(os.Args, " ")
+	}
+
+	// Required is a map of field name to required preceding field name.
+	// This is used to ensure even empty (and otherwise omitted) fields can be
+	// included for documentation purposes.
+	required := map[string]string{
+		"resources":  "",
+		"parameters": "resources",
+		"scenarios":  "parameters",
+		"objectives": "scenarios",
+	}
```

Comment:
```suggestion
	required := map[string]string{
  		"resources":  "",
  		"parameters": "resources",
  		"scenarios":  "parameters",
  		"objectives": "scenarios",
  		"catchall": "objectives",
  	}
```
not that I was trying to break things, but if I run `redskyctl generate application --objectives ""`, we don't get any of the required output flagged. 
If we were to add in a catchall key with a value of `objectives`, I get the expected output
```

Reply:
Good catch, I used `""` as a key and added a call to `missingRequiredContent` after the for loop. I'm not sure I follow the GitHub suggestion thing above, but I didn't see the part where missing content was called with the "catchall".

Reply:
I thought the suggestion would be good to highlight the diff; we basically add an additional kv pair to the required list with a value of 'objectives'.  ex --
```
"catchall": "objectives"
```
as we traverse through the generated yaml, the missingRequiredContent handles pruning this map, so having a `catchall` ( or whatever value that is a non top level application key should work ). Since we point to objectives, we traverse through the whole map, so this extra KV pair triggers the missingRequiredContent to fill in the rest.